File: problem-solving/_lc-62-medium-unique-paths/2-bottom-up.py
"""
https://leetcode.com/problems/unique-paths/

Complexity
----------
Time: O(m*n)
Space: O(m*n)
"""

class Solution:
    def uniquePaths(self, m: int, n: int) -> int:
        dp = [[0] * (n + 1) for _ in range(m + 1)]
        for r in range(1, m + 1):
            for c in range(1, n + 1):
                if r == 1 or c == 1:
                    dp[r][c] = 1
                    continue
                dp[r][c] = dp[r - 1][c] + dp[r][c - 1]
        return dp[m][n]

File: problem-solving/lc-96-medium-unique-binary-search-trees/2-bottom-up.go
/*
https://leetcode.com/problems/unique-binary-search-trees/

Complexity
----------
Time: O(n^2)
Space: O(n)

Reference: https://leetcode.com/problems/unique-binary-search-trees/solutions/1565543/c-python-5-easy-solutions-w-explanation-optimization-from-brute-force-to-dp-to-catalan-o-n/
*/

func numTrees(n int) int {
    dp := make([]int, n + 1)
    dp[0], dp[1] = 1, 1

    for i := 2; i <= n; i++ {
        for j := 1; j <= i; j++ {
            dp[i] += dp[j - 1] * dp[i - j]
        }
    }
    return dp[n]
}
